README.md
# Face Recognition
## Siamese Neural Network

### Anchor images , negative images and position images

we use Siamese neural network for compare anchor images with negative images and position images 

position images includes images of that person you want recognize 

negative images includes images of another persons 

we use this collection for labeling and training 

## implement Siamese Neural Network

The following article gives you a good idea of this neural network 

https://www.cs.cmu.edu/~rsalakhu/papers/oneshot1.pdf


### Setup commands

## Downdload lfw.tgz

lfw is Labeled Faces in the Wild, a database of face photographs designed for studying the problem of unconstrained face recognition.
so dowload dataset from flolowing link:

http://vis-www.cs.umass.edu/lfw/

we use this dataset as negative images

* save in root of project

## Install Dependencies

pip install tensorflow==2.4.1 tensorflow-gpu==2.4.1 opencv-python matplotlib


## Run create_files.py

run this script for create anchor , negative and position directories and also move lfw dataset to negative directory 

python create_files.py

## Run create_ancher_position.py

for collecting anchor images and position images with webcam ,  run create_anchor_position.py 

- press 'p' when you want save images for position dataset
- press 'a' when you want save anchor dataset  

python create_ancher_position.py

## Run train script

python train.py

- after training this will save model 

## Run App with siamesemodel.h5

we use siamesemodel.h5 that we've saved model in training process

python app.py

CHANGELOG.md
# Face Recognition
## [1.0.0] - 18 Mar 2022
### Added
- install dependencies
- download lfw dataset , create directories (anchor , negative , position)
- copy images from lfw to data/negative
- save anchor and position dataset with webcam (opencv)
- train model
- save model 
- use model for Face Recognition

app.py
import cv2
import os
import numpy as np
from tensorflow.keras.layers import Layer
import tensorflow as tf 





def preprocess(file_path):
    byte_img = tf.io.read_file(file_path)
    img = tf.io.decode_jpeg(byte_img)
    img = tf.image.resize(img,(100 , 100))
    img = img / 255.0
    return img



class L1Dist(Layer):
    def __init__(self,**kwargs):
        super().__init__()
        
    def call(self , input_embedding, validation_embedding):
        return tf.math.abs( input_embedding - validation_embedding)
    
    
# Loss function

binary_cross_loss = tf.losses.BinaryCrossentropy()

  








model = tf.keras.models.load_model('siamesemodel.h5' , custom_objects={'L1Dist':L1Dist , 'BinaryCrossentropy':binary_cross_loss})


def verify(model , detection_threshold , verification_threshold):
    results = []
    for image in os.listdir(os.path.join('application_data', 'verification_images')):
        input_img = preprocess(os.path.join('application_data' , 'input_image' , 'input_image.jpg'))
        if image[-4:] != ".jpg" : continue 
        validation_img = preprocess(os.path.join('application_data' , 'verification_images' , image))
        
        result = model.predict(list(np.expand_dims([input_img , validation_img] , axis=1)))
        results.append(result)
    detection = np.sum(np.array(results) > detection_threshold)  
    
    verification = detection / len(os.listdir(os.path.join('application_data' , 'verification_images'))) 
    
    verified = verification > verification_threshold
    
    return results , verified
    
     
cap = cv2.VideoCapture(0)
result = False
while cap.isOpened():
    ret , frame = cap.read()
    # CUT DOWN FRAME to 250 * 250 pixels
    frame = frame[:250 , :250 , :]
    if result :
        cv2.putText(frame , "verified" , (20 , 50) , cv2.FONT_HERSHEY_COMPLEX , 1 , 255 , 1)
    else:
        cv2.putText(frame , "Not verified" , (20 , 50) , cv2.FONT_HERSHEY_COMPLEX , 1 , 255 , 1)
        
    cv2.imshow('Verification' , frame)
    
    if cv2.waitKey(1) & 0XFF == ord('v'):
        #save input image
        cv2.imwrite(os.path.join('application_data' , 'input_image' , 'input_image.jpg') , frame)
        # run verification
        results , verified = verify(model , 0.5 , 0.5)
        result = verified
        
    if cv2.waitKey(2) & 0XFF == ord('q'):
        break
    
cap.release()
cv2.destroyAllWindows()




        
         
        


 
    
    




    
    
    



create_files.py
import cv2
import os
import random
import numpy as np
from matplotlib import pyplot as plt


# import tensorflow dependencies - functional apis   

from tensorflow.keras.models import Model
from tensorflow.keras.layers import Layer , Conv2D  , Dense , MaxPooling2D , Input , Flatten
import tensorflow as tf 

# Import uuid library for generate unique image names

import uuid 

# setup path

POS_PATH = os.path.join('data' , 'position')

NEG_PATH = os.path.join('data' , 'negative')

ANC_PATH = os.path.join('data' , 'anchor')

# create directories
os.makedirs(POS_PATH)
os.makedirs(NEG_PATH)
os.makedirs(ANC_PATH)

# move lfw images to data/negative dir

for directory in os.listdir('lfw'):
    for file in os.listdir(os.path.join('lfw' , directory)):
        Ex_PATH = os.path.join('lfw' , directory, file)
        NEW_PATH = os.path.join(NEG_PATH , file)
        os.replace(Ex_PATH, NEW_PATH)


  

create_anchor_position.py
import cv2
import os
import random
import numpy as np
from matplotlib import pyplot as plt


# import tensorflow dependencies - functional apis   

from tensorflow.keras.models import Model
from tensorflow.keras.layers import Layer , Conv2D  , Dense , MaxPooling2D , Input , Flatten
import tensorflow as tf 

# Import uuid library for generate unique image names

import uuid 

# setup path

POS_PATH = os.path.join('data' , 'position')

NEG_PATH = os.path.join('data' , 'negative')

ANC_PATH = os.path.join('data' , 'anchor')




# establish a connection to the webcam

cap = cv2.VideoCapture(0)

while cap.isOpened():
    ret , frame = cap.read()
    # CUT DOWN FRAME to 250 * 250 pixels
    frame = frame[:250 , :250 , :]
    cv2.imshow('Image collection' , frame)
    # collect anchors 
    if cv2.waitKey(1) & 0XFF == ord('p'):
        img_name = os.path.join(POS_PATH , '{}.jpg'.format(uuid.uuid1()))
        cv2.imwrite(img_name , frame)
    if cv2.waitKey(2) & 0XFF == ord('a'):
        img_name = os.path.join(ANC_PATH , '{}.jpg'.format(uuid.uuid1()))
        cv2.imwrite(img_name , frame)
    # collect positives
    # show images
    if cv2.waitKey(3) & 0XFF == ord('q'):
        break
    
cap.release()
cv2.destroyAllWindows()


train.py
import cv2
import os
import random
import numpy as np
from matplotlib import pyplot as plt


# import tensorflow dependencies - functional apis   

from tensorflow.keras.models import Model
from tensorflow.keras.layers import Layer , Conv2D  , Dense , MaxPooling2D , Input , Flatten
import tensorflow as tf 
# import metrix evaluations
from tensorflow.keras.metrics import Precision , Recall


# Import uuid library for generate unique image names

import uuid 

# setup path

POS_PATH = os.path.join('data' , 'position')

NEG_PATH = os.path.join('data' , 'negative')

ANC_PATH = os.path.join('data' , 'anchor')



anchor = tf.data.Dataset.list_files(ANC_PATH+'/*.jpg').take(300)

positive = tf.data.Dataset.list_files(POS_PATH+'/*.jpg').take(300)

negative = tf.data.Dataset.list_files(NEG_PATH+'/*.jpg').take(300)
  


def preprocess(file_path):
    byte_img = tf.io.read_file(file_path)
    img = tf.io.decode_jpeg(byte_img)
    img = tf.image.resize(img,(100 , 100))
    img = img / 255.0
    return img


positives = tf.data.Dataset.zip((anchor , positive , tf.data.Dataset.from_tensor_slices(tf.ones(len(anchor)))))
negatives = tf.data.Dataset.zip((anchor , negative , tf.data.Dataset.from_tensor_slices(tf.zeros(len(anchor)))))
data = positives.concatenate(negatives)

def preprocess_twin(input_img , validation_img , label):
    return(preprocess(input_img), preprocess(validation_img),label)

#build dataloader pipeline
data = data.map(preprocess_twin)
data = data.cache()
data = data.shuffle(buffer_size=1024)

# training partion
train_data = data.take(round(len(data)*.7))
train_data = train_data.batch(16)
train_data = train_data.prefetch(8)

# testing partion
test_data = data.skip(round(len(data)*.7))
test_data = test_data.take(round(len(data)*.3))
test_data = test_data.batch(16)
test_data = test_data.prefetch(8) 



def make_embedding():
    inp = Input(shape=(100 , 100 , 3) , name='input_image')
    # First block
    c1 = Conv2D(64 , (10 , 10) , activation='relu')(inp)
    m1 = MaxPooling2D(64 , (2 , 2) , padding='same')(c1)
    # Seconde block
    c2 = Conv2D(128 , (7 , 7) , activation='relu')(m1)
    m2 = MaxPooling2D(64 , (2 , 2) , padding='same')(c2)
    # Third block
    c3 = Conv2D(128 , (4 , 4) , activation='relu')(m2)
    m4 = MaxPooling2D(64 , (2 , 2) , padding='same')(c3)
    # finall embedding layer
    c4 = Conv2D(256 , (4 , 4) , activation='relu')(m4)
    f1 = Flatten()(c4)
    d1 = Dense(4094 , activation='sigmoid')(f1)
    
    
    return Model(inputs=[inp],outputs=[d1],name='embedding')

embedding = make_embedding()

class L1Dist(Layer):
    def __init__(self,**kwargs):
        super().__init__()
        
    def call(self , input_embedding, validation_embedding):
        return tf.math.abs( input_embedding - validation_embedding)
    
    
    
def make_siamese_model():
    # anchor image input in the network
    input_image = Input(name='input_img' , shape=(100 , 100 , 3))
    # validation image in the network
    validation_image = Input(name='validation_img' , shape=(100 , 100 , 3))
    
    #combine siamese distance components
    siamese_layer = L1Dist()
    
    siamese_layer._name = 'distance'
    distances = siamese_layer(embedding(input_image), embedding(validation_image))
    
    
    #classification layer
    
    classifier = Dense(1 , activation='sigmoid')(distances) 
    
    return Model(inputs=[input_image, validation_image] , outputs=[classifier] , name= 'SiameseNetwork')

# create model
siamese_model = make_siamese_model()
# Loss function

binary_cross_loss = tf.losses.BinaryCrossentropy()


# Optimization function
opt = tf.keras.optimizers.Adam(1e-4)


# checkpoint 

checkpoint_dir = './training_checkpoints'
checkpoint_prefix = os.path.join(checkpoint_dir , 'ckpt')
checkpoint = tf.train.Checkpoint(opt=opt , siamese_model=siamese_model)


@tf.function
def train_step(batch):
    
    with tf.GradientTape() as tape:
        
        #get anchor and pos/neg image 
        
        X = batch[:2]
        #get label
        y = batch[2]
        
        # forward pass
        yhat = siamese_model(X , training=True)
        #calculate loss 
        loss = binary_cross_loss(y , yhat)
    print(loss)
        # calculate gradients
    grad = tape.gradient(loss , siamese_model.trainable_variables)
    
    # calculate updated weights and apply to siamese models
    
    opt.apply_gradients(zip(grad , siamese_model.trainable_variables))
    
    return loss


def train(data , EPOCHS):
    # loop through epochs
    
    for epoch in range(1 , EPOCHS+1):
        print('\n Epoch {}/{}'.format(epoch, EPOCHS))
        progbar = tf.keras.utils.Progbar(len(data))
        
        # loop through each batch
        
        for idx , batch in enumerate(train_data):
            train_step(batch)
            progbar.update(idx+1)
            
        if epoch % 10 == 0:
            checkpoint.save(file_prefix = checkpoint_prefix)
            
            
EPOCHS = 50
train(train_data , EPOCHS)


            
            
    
    
    
# TEST        
    
    
# make predictions
test_input , test_val , y_true = test_data.as_numpy_iterator().next()
y_hat = siamese_model([test_input , test_val])

m = Recall()
m.update_state(y_true , y_hat)
print(m.result().numpy())


p = Precision()
p.update_state(y_true , y_hat)
print(p.result().numpy())




# save weights
siamese_model.save('siamesemodel.h5')


         
        


 
    
    




    
    
    

